### PR TITLE
Make the cached mirror methods invisible

### LINKED ISSUE #13815: synthetic Mirror methods become visible when from classpath dependency
A companion object of a case class from a classpath dependency will have synthetic Mirror methods visible, e.g. `fromProduct`. However these methods are deliberately hidden when compiling that original case class until after `posttyper`. This means there is discrepancy between library dependencies and source dependencies.

## Compiler version

3.1.0

## Minimized code

```Scala
case class Box(i: Int, s: String)
val box = Box.fromProduct((1, "a"))
```

## Output

```scala
1 |val box = Box.fromProduct((1, "a"))
  |          ^^^^^^^^^^^^^^^
  |          value fromProduct is not a member of object Box
```

## Expectation

either
- `fromProduct` is not available from dependencies
- `fromProduct` is visible at typer

## Notes

If I instead compile `Box` previously and load it on the classpath then `Box.fromProduct((1, "a"))` is allowed in a new file.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #13815

### CODE REVIEW

**Review by bishabosha — CHANGES_REQUESTED** [review]
the ordinal method should also be made invisible:
```scala
// Shape_1.scala
sealed trait Shape

object Shape:
  case class Circle(r: Int) extends Shape
  case class Square(width: Int) extends Shape
```

```scala
// Test_2.scala
val circleOrdinal = Shape.ordinal(Shape.Circle(r = 23))
```

**Review by bishabosha — CHANGES_REQUESTED** [review]
Just thinking now that the `MirroredMonoType` is also visible too

**Comment by odersky** [comment]
Right. Fix is forthcoming.

**Comment by bishabosha** [comment]
however I have thought of something else, which is that the companion will still extend `Mirror.Sum` or `Mirror.Product` - so you can widen the type to call the method:
```scala
4 |val circleOrdinal = (Shape: deriving.Mirror.Sum).ordinal(Shape.Circle(r = 23)) // error
  |                                                         ^^^^^^^^^^^^^^^^^^^^
  |              Found:    Shape.Circle
  |              Required: ?1.MirroredMonoType
  |
  |              where:    ?1 is an unknown value of type deriving.Mirror.Sum
```
could there be a way to make a parent invisible?

File: tests/neg/i13815.check [hunk 1]
```diff
@@ -3,7 +3,14 @@
 1 |val box = Box.fromProduct((1, "a"))  // error
   |          ^^^^^^^^^^^^^^^
   |          value fromProduct is not a member of object Box
--- [E008] Not Found Error: tests/neg/i13815/Test_2.scala:2:26 ----------------------------------------------------------
+-- [E007] Type Mismatch Error: tests/neg/i13815/Test_2.scala:2:46 ------------------------------------------------------
 2 |val circleOrdinal = Shape.ordinal(Shape.Circle(r = 23)) // error
-  |                    ^^^^^^^^^^^^^
-  |                    value ordinal is not a member of object Shape
+  |                                  ^^^^^^^^^^^^^^^^^^^^
+  |                                  Found:    Shape.Circle
```
Comment (bishabosha) [inline tests/neg/i13815.check:9]: this seems strange that making `MirroredMonoType` invisible now makes `ordinal` visible again

**Comment by bishabosha** [comment]
maybe this needs to be solved on the unpickling side - ignore synthethic mirror stuff from tasty, then in posttyper re-add them

**Comment by odersky** [comment]
> maybe this needs to be solved on the unpickling side - ignore the mirror parent in tasty (if it has invisible/synthetic mirror methods), then in posttyper re-add the parent

I believe that would be a hole of complexity.

**Comment by odersky** [comment]
> could there be a way to make a parent invisible?

I don't think so. I don't really want to sink more time into this. It is what it is.

**Comment by odersky** [comment]
I think we should close this. I don't see a good way to address the issue currently.

**Pull request closed without merging**